src/dust.ts

```diff
@@ -0,0 +1,196 @@
+            this._particles.push(new Particle(model, step));
+        }
+    }
+
+    /**
+     * Returns the number of state elements per particle
+     */
+    public nState(): number {
+        return this._particles[0].size;
+    }
+
+    /**
+     * Returns the number of particles
+     */
+    public nParticles(): number {
+        return this._particles.length;
+    }
+
+    /**
+     * The current step
+     */
+    public step(): number {
+        return this._particles[0].step;
+    }
+
+    /**
+     * Returns information about how the state is packed. This is
+     * useful for computing indexes and extracting state.
+     */
+    public info(): DustModelInfo {
+        return this._particles[0].info();
+    }
+
+    /**
+     * Set new parameters into the model
+     *
+     * @param pars: New parameters
+     *
+     * @param setInitialState If `true`, then changing the parameters
+     * also updates the initial conditions to use those from the
+     * model's `initial()` method. This uses the current step of the
+     * model, so you may want to use {@link Dust.setStep} first.
+     */
+    public setPars(pars: Pars, setInitialState: boolean): void {
+        const step = this.step();
+        const nState = this.nState();
+        const model = new this._Model(base, pars);
+        if (model.size() !== this.nState()) {
+            throw Error(`Particle produced unexpected state size`);
+        }
+        for (let i = 0; i < this._particles.length; ++i) {
+            const state =
+                setInitialState ? undefined : this._particles[i].state();
+            this._particles[i] = new Particle(model, step, state);
+        }
+    }
+
+    /**
+     * Change the step in the model, without updating state
+     *
+     * @param step New step
+     */
+    public setStep(step: number): void {
+        this._particles.forEach((p) => {
+            p.step = step;
+        });
+    }
+
+    /**
+     * Change the model state
+     *
+     * @param state A 2d-matrix of state; this inteface may change.
+     */
+    public setState(state: number[][]): void {
+        this.checkState(state);
+        for (let i = 0; i < this.nParticles(); ++i) {
+            this._particles[i].setState(state[i]);
+        }
+    }
+
+    /**
+     * Run the model up to some step
+     *
+     * @param stepEnd The step to run the model to
+     */
+    public run(stepEnd: number): void {
+        this._particles.forEach((p: Particle) => p.run(stepEnd, this._random));
+    }
+
+    /**
+     * Run the model and collect state at a series of times
+     *
+     * @param stepEnd An array of steps to collect state at, must be
+     * increasing. The model will stop running at the final element.
+     *
+     * @param index An index to use to filter state, or `null` to
+     * return the full state, see {@link Dust.state}
+     */
+    public simulate(stepEnd: number[], index: number[] | null): DustStateTime {
```

Comment:
```suggestion
    public simulate(collectSteps: number[], index: number[] | null): DustStateTime {
```
Would `collectSteps` be a more accurate name for this parameter since it include intermediate steps, not just the last one required?

Reply:
it might be, though this reflects the name used in dust (https://github.com/mrc-ide/dust/blob/master/R/dust_generator.R#L159) so I'd rather keep it. The other reason it's useful is that it reminds me that collection happens once the step is completed (i.e. at the end of the step not the start)